src/models/transformer/decoder.rs

```diff
@@ -0,0 +1,184 @@
+    n_hidden_layers: usize,
+    output_layer_norm: Box<dyn BuildModule>,
+}
+
+impl TransformerDecoderConfig {
+    /// Decoder embeddings.
+    ///
+    /// Default: `TransformerEmbeddingsConfig::default()`
+    pub fn embeddings(mut self, embeddings: TransformerEmbeddingsConfig) -> Self {
+        self.embeddings = embeddings;
+        self
+    }
+
+    /// Decoder layer.
+    ///
+    /// Default: `TransformerLayerConfig::default()`
+    pub fn layer(mut self, layer: Box<dyn BuildDecoderLayer<Cache = KeyValueCache>>) -> Self {
+        self.layer = layer;
+        self
+    }
+
+    /// Number of hidden layers.
+    ///
+    /// Default: `12`
+    pub fn n_hidden_layers(mut self, n_hidden_layers: usize) -> Self {
+        self.n_hidden_layers = n_hidden_layers;
+        self
+    }
+
+    /// Output layer normalization module.
+    ///
+    /// Default: `Identity`
+    pub fn output_layer_norm(mut self, output_layer_norm: Box<dyn BuildModule>) -> Self {
+        self.output_layer_norm = output_layer_norm;
+        self
+    }
+}
+
+impl BuildDecoder for TransformerDecoderConfig {
+    type Decoder = TransformerDecoder;
+
+    fn build(&self, vb: VarBuilder) -> Result<TransformerDecoder, BoxedError> {
+        let embeddings = self
+            .embeddings
+            .build(vb.push_prefix("embeddings"))
+            .context(TransformerEmbeddingsSnafu)?;
+
+        let layers = (0..self.n_hidden_layers)
+            .map(|n| {
+                self.layer
+                    .build_decoder_layer(vb.push_prefix(format!("layer_{n}")))
+                    .context(BuildTransformerLayerSnafu)
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+
+        Ok(TransformerDecoder {
+            embeddings,
+            layers,
+            output_layer_norm: self
+                .output_layer_norm
+                .build(vb.push_prefix("output_layer_norm"))
+                .context(BuildLayerNormSnafu)?,
+        })
+    }
+}
+
+impl Default for TransformerDecoderConfig {
+    fn default() -> Self {
+        Self {
+            embeddings: TransformerEmbeddingsConfig::default(),
+            layer: Box::<TransformerLayerConfig>::default(),
+            n_hidden_layers: 12,
+            output_layer_norm: Box::new(Identity),
+        }
+    }
+}
+
+/// Transformer decoder errors.
+#[derive(Debug, Snafu)]
+pub enum TransformerDecoderError {
+    #[snafu(display("Cannot build layer norm"))]
+    BuildLayerNorm { source: BoxedError },
+
+    #[snafu(display("Cannot build transformer layer"))]
+    BuildTransformerLayer { source: BoxedError },
+
+    #[snafu(display("Cannot construct or apply embeddings"))]
+    Embedding { source: TransformerEmbeddingsError },
+
+    #[snafu(display("Cannot construct or apply layer norm"))]
+    LayerNorm { source: candle_core::Error },
+
+    #[snafu(display("Cannot construct or apply embeddings"))]
+    TransformerEmbeddings { source: TransformerEmbeddingsError },
```

Comment:
I think we can rename this to `BuildTransformerEmbeddings` to avoid confusion?

Reply:
Indeed, fixed!